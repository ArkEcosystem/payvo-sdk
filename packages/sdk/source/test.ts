/* istanbul ignore file */

import { randomUUID } from "crypto";
import { ConfigKey, ConfigRepository } from "./coins";
import { Container, BindingType } from "./ioc";
import { NetworkManifest } from "./networks";
import { BigNumberService } from "./services";

const createContainer = ({
	config,
	httpClient,
	manifest,
	meta,
}: {
	config?: ConfigRepository;
	httpClient: any;
	manifest: NetworkManifest;
	meta?: any;
}): Container => {
	config ??= new ConfigRepository({ network: manifest.id, httpClient });

	config.set(ConfigKey.Network, manifest);

	if (meta) {
		for (const [key, value] of Object.entries(meta)) {
			config.set(key, value);
		}
	}

	const container = new Container();
	container.constant(BindingType.ConfigRepository, config);
	container.constant(BindingType.HttpClient, httpClient);
	container.singleton(BindingType.BigNumberService, BigNumberService);

	return container;
};

export const createService = <T = any>({
	config,
	httpClient,
	manifest,
	meta,
	predicate,
	service,
}: {
	config?: ConfigRepository;
	httpClient: any;
	manifest: NetworkManifest;
	meta?: any;
	predicate: any;
	service: any;
}): T => {
	const container = createContainer({
		config,
		httpClient,
		manifest,
		meta,
	});

	if (predicate) {
		predicate(container);
	}

	return container.resolve(service);
};

export const createServiceAsync = async <T = any>({
	config,
	httpClient,
	manifest,
	meta,
	predicate,
	service,
}: {
	config?: ConfigRepository;
	httpClient: any;
	manifest: NetworkManifest;
	meta?: any;
	predicate: any;
	service: any;
}): Promise<T> => {
	const container = createContainer({
		config,
		httpClient,
		manifest,
		meta,
	});

	if (predicate) {
		await predicate(container);
	}

	const uuid: string = randomUUID();

	container.singleton(uuid, service);

	return container.get(uuid);
};
