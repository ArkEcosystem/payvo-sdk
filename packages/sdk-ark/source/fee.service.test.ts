import "jest-extended";

import nock from "nock";

import { createService, require } from "../test/mocking";
import { FeeService } from "./fee.service";

const matchSnapshot = (transaction): void =>
	expect({
		min: transaction.min.toString(),
		avg: transaction.avg.toString(),
		max: transaction.max.toString(),
		static: transaction.static.toString(),
	}).toMatchSnapshot();

afterEach(() => nock.cleanAll());

beforeAll(() => nock.disableNetConnect());

describe("FeeService", () => {
	it("should get the fees for ARK", async () => {
		nock(/.+/)
			.get("/api/node/fees")
			.reply(200, await require(`../test/fixtures/client/feesByNode.json`))
			.get("/api/transactions/fees")
			.reply(200, await require(`../test/fixtures/client/feesByType.json`));

		const result = await (await createService(FeeService, "ark.devnet")).all();

		expect(result).toContainAllKeys([
			"transfer",
			"secondSignature",
			"delegateRegistration",
			"vote",
			"multiSignature",
			"ipfs",
			"multiPayment",
			"delegateResignation",
			"htlcLock",
			"htlcClaim",
			"htlcRefund",
		]);

		matchSnapshot(result.transfer);
		matchSnapshot(result.secondSignature);
		matchSnapshot(result.delegateRegistration);
		matchSnapshot(result.vote);
		matchSnapshot(result.multiSignature);
		matchSnapshot(result.ipfs);
		matchSnapshot(result.multiPayment);
		matchSnapshot(result.delegateResignation);
		matchSnapshot(result.htlcLock);
		matchSnapshot(result.htlcClaim);
		matchSnapshot(result.htlcRefund);
	});

	it("should get the fees for BIND", async () => {
		nock(/.+/)
			.get("/api/node/fees")
			.reply(200, await require(`../test/fixtures/client/feesByNode-bind.json`))
			.get("/api/transactions/fees")
			.reply(200, await require(`../test/fixtures/client/feesByType-bind.json`));

		const result = await (await createService(FeeService, "bind.testnet")).all();

		expect(result).toContainAllKeys([
			"transfer",
			"secondSignature",
			"delegateRegistration",
			"vote",
			"multiSignature",
			"ipfs",
			"multiPayment",
			"delegateResignation",
			"htlcLock",
			"htlcClaim",
			"htlcRefund",
		]);

		matchSnapshot(result.transfer);
		matchSnapshot(result.secondSignature);
		matchSnapshot(result.delegateRegistration);
		matchSnapshot(result.vote);
		matchSnapshot(result.multiSignature);
		matchSnapshot(result.ipfs);
		matchSnapshot(result.multiPayment);
		matchSnapshot(result.delegateResignation);
		matchSnapshot(result.htlcLock);
		matchSnapshot(result.htlcClaim);
		matchSnapshot(result.htlcRefund);
	});

	it("should calculate the fees for ARK", async () => {
		nock(/.+/)
			.get("/api/node/fees")
			.reply(200, await require(`../test/fixtures/client/feesByNode.json`))
			.get("/api/transactions/fees")
			.reply(200, await require(`../test/fixtures/client/feesByType.json`))
			.persist();

		const a = await (
			await createService(FeeService, "ark.devnet")
		).calculate({
			type: 4,
			asset: { multiSignature: { publicKeys: ["a", "b", "c"] } },
		});
		const b = await (await createService(FeeService, "ark.devnet")).calculate({ type: 1 });

		expect(a.toHuman()).toBe(20);
		expect(b.toHuman()).toBe(0);
	});
});
