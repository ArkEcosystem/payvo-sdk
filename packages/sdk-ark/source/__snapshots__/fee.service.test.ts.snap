// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeeService should get the fees for ARK 1`] = `
Object {
  "avg": "9878740",
  "isDynamic": true,
  "max": "10000000",
  "min": "3627425",
  "static": "10000000",
}
`;

exports[`FeeService should get the fees for ARK 2`] = `
Object {
  "avg": "0",
  "isDynamic": true,
  "max": "500000000",
  "min": "0",
  "static": "500000000",
}
`;

exports[`FeeService should get the fees for ARK 3`] = `
Object {
  "avg": "0",
  "isDynamic": true,
  "max": "2500000000",
  "min": "0",
  "static": "2500000000",
}
`;

exports[`FeeService should get the fees for ARK 4`] = `
Object {
  "avg": "52995956",
  "isDynamic": true,
  "max": "100000000",
  "min": "1000000",
  "static": "100000000",
}
`;

exports[`FeeService should get the fees for ARK 5`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "500000000",
  "min": "0",
  "static": "500000000",
}
`;

exports[`FeeService should get the fees for ARK 6`] = `
Object {
  "avg": "62128036",
  "isDynamic": true,
  "max": "500000000",
  "min": "10000000",
  "static": "500000000",
}
`;

exports[`FeeService should get the fees for ARK 7`] = `
Object {
  "avg": "10000000",
  "isDynamic": true,
  "max": "10000000",
  "min": "10000000",
  "static": "10000000",
}
`;

exports[`FeeService should get the fees for ARK 8`] = `
Object {
  "avg": "0",
  "isDynamic": true,
  "max": "2500000000",
  "min": "0",
  "static": "2500000000",
}
`;

exports[`FeeService should get the fees for ARK 9`] = `
Object {
  "avg": "10000000",
  "isDynamic": true,
  "max": "10000000",
  "min": "10000000",
  "static": "10000000",
}
`;

exports[`FeeService should get the fees for ARK 10`] = `
Object {
  "avg": "0",
  "isDynamic": true,
  "max": "0",
  "min": "0",
  "static": "0",
}
`;

exports[`FeeService should get the fees for ARK 11`] = `
Object {
  "avg": "0",
  "isDynamic": true,
  "max": "0",
  "min": "0",
  "static": "0",
}
`;

exports[`FeeService should get the fees for BIND 1`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "10000000",
  "min": "0",
  "static": "10000000",
}
`;

exports[`FeeService should get the fees for BIND 2`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "50000000",
  "min": "0",
  "static": "50000000",
}
`;

exports[`FeeService should get the fees for BIND 3`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "25000000000",
  "min": "0",
  "static": "25000000000",
}
`;

exports[`FeeService should get the fees for BIND 4`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "100000000",
  "min": "0",
  "static": "100000000",
}
`;

exports[`FeeService should get the fees for BIND 5`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "50000000",
  "min": "0",
  "static": "50000000",
}
`;

exports[`FeeService should get the fees for BIND 6`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "0",
  "min": "0",
  "static": "0",
}
`;

exports[`FeeService should get the fees for BIND 7`] = `
Object {
  "avg": "25000000",
  "isDynamic": false,
  "max": "25000000",
  "min": "25000000",
  "static": "25000000",
}
`;

exports[`FeeService should get the fees for BIND 8`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "1000000000",
  "min": "0",
  "static": "1000000000",
}
`;

exports[`FeeService should get the fees for BIND 9`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "0",
  "min": "0",
  "static": "0",
}
`;

exports[`FeeService should get the fees for BIND 10`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "0",
  "min": "0",
  "static": "0",
}
`;

exports[`FeeService should get the fees for BIND 11`] = `
Object {
  "avg": "0",
  "isDynamic": false,
  "max": "0",
  "min": "0",
  "static": "0",
}
`;
